### ISSUE TITLE
Losing in leader election clears need_rewind flag

### ISSUE THREAD

**Issue by ants** [zalando/patroni#347]
``process_unhealthy_cluster()` does:

``` python
    need_rewind = bool(self.cluster.failover) or self.patroni.nofailover
```

and then `Postgresql.follow()` does:

``` python
if need_rewind is not None:
    self._need_rewind = need_rewind
```

This looks incorrect to me. I assume tacking on an `or None` default onto there is what was intended. I'll continue my refactoring based on this assumption, but wanted to double check as the need_rewind logic is hard to follow.

**Comment by ants** [comment]
Another thing in the same area of code. Situation - 3 node cluster (A,B,C), Patroni is paused, there is a manual failover key for A->B, leader A has just demoted itself. B and C are racing to run run_cycle.

1. If B runs before C, then B acquires leader lock, promoets itself. C does not set the rewind flag, starts following B.
2. If C runs before B, then C sets the rewind flag, follow() return 'no action'. Now B runs. C runs next cycle, then next cycle C will do a rewind and start following B.
3. LIke case 2, but B does not manage to run between the two C cycles. Now C will call Postgresql.follow, where the rewind flag is cleared because there is no leader, and C is restarted as a standby without primary_conninfo. When B now runs, C will not rewind from it.

That seems somewhat inconsistent. Why is the rewind flag being cleared if follow gets called without a leader?

**Comment by ants** [comment]
I managed to forget that I already removed the rewind flag clearing in 7ea6f3b to be included in #295.

**Comment by CyberDem0n** [comment]
Actually I have an Idea how to make pg_rewind is needed detection bulletproof:
```console

# get timeline and xlogpos of the follower (local node)
$ psql "host=localhost dbname=postgres replication=database user=replicator port=5432" -c "IDENTIFY_SYSTEM;"
      systemid       │ timeline │  xlogpos  │  dbname  
─────────────────────┼──────────┼───────────┼──────────
 6346153464845775387 │        1 │ 0/30001B0 │ postgres
(1 row)

# get timeline of the master
$ psql "host=localhost dbname=postgres replication=database user=replicator port=5433" -c "IDENTIFY_SYSTEM;"
      systemid       │ timeline │  xlogpos  │  dbname  
─────────────────────┼──────────┼───────────┼──────────
 6346153464845775387 │        2 │ 0/3000330 │ postgres
(1 row)

# get timeline history from the master
$ psql "host=localhost dbname=postgres replication=database user=replicator port=5433" -c "TIMELINE_HISTORY 2;"
     filename     │                       content                        
──────────────────┼──────────────────────────────────────────────────────
 00000002.history │ 1       0/3000140       no recovery target specified↵
                  │ 
(1 row)

```

What we can see here:
 - the new master is on timeline 2 and timeline was created when lsn was '0/3000140'.
 - local postgres is on timeline 1 but its lsn is '0/30001B0'

'0/30001B0' is ahead of  '0/3000140', that mean we have to run 'pg_rewind'

What do you think?

**Comment by ants** [comment]
Using a replication connection to fetch timeline information is a great idea! I was postponing work in this area thinking getting the timeline information would be something much more complex. Managing a boolean need_rewind flag is complicated, error prone and by the looks of it, fundamentally flawed.

The main issue I see, if the server was not shut down properly we need to first replay all locally available WAL before we know what the real WAL endpoint is. So the general idea would be - when recovering, bring the database up, do the timeline check, if we seem to have forked off somewhere, stop, rewind and restart.

**Comment by CyberDem0n** [comment]
I believe https://github.com/zalando/patroni/pull/417 solved it
